### ISSUE TITLE
I cant get my firebase data

### ISSUE THREAD

**Issue by imepnja** [flutter/flutter#57442]
Hey, I'm trying to check if a user is already registered or if i need to create a new account.
The idea was that i wanted to get a list of users and check if my new email matches up. If not create a new user 
I have a `DatabaseService` class which I use to get my data, and a `AuthService` which i use to perform login and registration.
I my `DatabaseService` class I have a function to return a list of `User` objects:

    List<User>  getUserList() {
    print(databaseReference);
    List<User> userList = [];
    databaseReference
        .collection("users")
        .getDocuments()
        .then((QuerySnapshot snapshot) {
          print('s');
      snapshot.documents.forEach((doc) {
        print(doc.documentID);
        userList.add(User(
          uid: doc.documentID,
          email: doc.data['email'],
          name: doc.data['name'],
          imgUrl: doc.data['imgUrl'],
        ));
        print(userList);
      });
    });
    return userList;
  }

This works fine when I call it from my `LoginPage`. I get a list with all my auhtorized users. But when i run it out of my  `loginIfRegistered` function in my `AuthService` class. I just get an empty list. this is the function:

      Future loginIfRegistered(String email, String password) async {
    print('CHECKING IF ALREADY REGISTERED');
    bool alreadyRegistered = false;
    
    List<User> _userList = DatabaseService().getUserList();

    for (var user in userList) {
      if (user.email == email) {
        alreadyRegistered = true;
      }
    }

    if (userList.length == 0) {
      alreadyRegistered = false;
    }

    
    if (alreadyRegistered) {
    // sign in  
    print('ALREADY REGISTERED');
      try {
        return await _auth.signInWithEmailAndPassword(
            email: email, password: password);
      } catch (e) {
        print(e.toString());
        return e;
      }
    } else {
      print('NOT REGISTERED');
      // register
      return null;
    }
  }

But when I pass the `userList` as a parameter into my function (`Future loginIfRegistered(String email, String password, List<User> userList) async {...}`), and initialize it on my `LoginPage` there is no problem.
I also tried getting the list at the top of my `AuthService` and it worked aswell, but i dont want it to be just updated when `AuthService` is initialized but rather when i hit the SignIn button. 

I also tried adding a `updateUserList` function in `AuthService`  which is called in `loginIfRegistered()` but that also doesent work.

**So why do I get an empty list if i call `getUserList()`?** 

To recap: I want to have a login page which also registers you if you don't already have an account.
When I press the LogIn Button the function `SigninIfRegisters` (in `AuthService`) is run which needs to know all users in my firestore Database (I made a collection for all my users, not out of Firebaseauth or something like that). Therefore i made a function in `DatabaseService` which gives me a `List<User>` this works fine if called out of the `LoginPage` but not out of any functions in `AuthService`. If i instantiate  `List<User>` not in any function in `AuthService` it also works.

Please someone explain Why????

**Comment by imepnja** [comment]
This is my `AuthService` file: 
  import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_test/models/user.dart';
import 'package:firebase_test/services/database.dart';

class AuthService {
  final FirebaseAuth _auth = FirebaseAuth.instance;

  String uid = '';

  Stream<FirebaseUser> get user {
    return _auth.onAuthStateChanged;
  }

  // Option One: works
  //
  // List<User> userList = DatabaseService().getUserList(); 
  //
  //My problem with this is if a user would create a account for example in a browser 
  // and has the app open at the same time. he wouldn be able to log in because it 
  //haden't refreshed the userList


  // Option Two:
  //
  // void updateUserList() {
  //   userList = DatabaseService().getUserList();
  // }
  //
  // this only works if I already have option one,
  // so i dont think this would actually change anything


  // option Tree: works
  // Future loginIfRegistered(String email, String password, List<User> userList) async {...}
  // 
  // but i dont want to update my userlist everytime i build my 
  // login page (aka everytime i enter a char in any of the forms)


  // what I want to work (and if you askme also shoul'd work)
  Future loginIfRegistered(String email, String password,) async {
    print('CHECKING IF ALREADY REGISTERED');
    bool alreadyRegistered = false;
    // updateUserList();

    // Running this line form anywhere ales works... but here I get an emply List
    List<User> userList = DatabaseService().getUserList();
    

    for (var user in userList) {
      if (user.email == email) {
        alreadyRegistered = true;
      }
    }

    if (userList.length == 0) {
      alreadyRegistered = false;
    }

    
    if (alreadyRegistered) {
    // sign in  
    print('ALREADY REGISTERED');
      try {
        return await _auth.signInWithEmailAndPassword(
            email: email, password: password);
      } catch (e) {
        print(e.toString());
        return e;
      }
    } else {
      print('NOT REGISTERED');
      // register
      return null;
    }
  }

  Future signUp({String email, String password, String username}) async {
    try {
      AuthResult result = await _auth.createUserWithEmailAndPassword(
          email: email, password: password);
      FirebaseUser user = result.user;

      DatabaseService().updateUserData(name: username, uid: user.uid);
      return user;
    } catch (e) {
      print(e.toString());
      return null;
    }
  }



  Future logout() {
    uid = '';

    _auth.signOut();
  }
}

Btw is there a way to avoid this weird formatting? I can't get it all to be `code`.

**Comment by imepnja** [comment]
[auth.zip](https://github.com/flutter/flutter/files/4640573/auth.zip)

**Comment by imepnja** [comment]
[database.zip](https://github.com/flutter/flutter/files/4640575/database.zip)

**Comment by imepnja** [comment]
[login.zip](https://github.com/flutter/flutter/files/4640576/login.zip)

**Comment by jonahwilliams** [comment]
The Github issue tracker is for bugs and features requests for the Flutter SDK. Help and support questions like these are better suited for Q&A sites like [StackOverflow](https://stackoverflow.com/questions/tagged/flutter) or on one of the flutter community sites at https://flutter.dev/community .
